Update 6-viewsets-and-routers.md (#8590)
encode/django-rest-framework: docs/tutorial/6-viewsets-and-routers.md

> # Tutorial 6: View sets & Routers

Reviewer: ```suggestion
# Tutorial 6: ViewSets & Routers
```

> REST framework includes an abstraction for dealing with view sets, that allows the developer to concentrate on modeling the state and interactions of the API, and leave the URL construction to be handled automatically, based on common conventions.

Reviewer: ```suggestion
REST framework includes an abstraction for dealing with `ViewSets`, that allows the developer to concentrate on modeling the state and interactions of the API, and leave the URL construction to be handled automatically, based on common conventions.
```

`ViewSet` classes are almost the same thing as `View` classes, except that they provide operations such as `retrieve`, or `update`, and not method handlers such as `get` or `put`.

A `ViewSet` class is only bound to a set of method handlers at the last moment, when it is instantiated into a set of views, typically by using a `Router` class which handles the complexities of defining the URL conf for you.

> ## Refactoring to use view sets

Reviewer: ```suggestion
## Refactoring to use ViewSets
```

Let's take our current set of views, and refactor them into view sets.

> First of all let's refactor our `UserList` and `UserDetail` classes into a single `UserViewSet` class.  We can remove the two view classes, and replace them with a single view set class:

Reviewer: ```suggestion
First of all let's refactor our `UserList` and `UserDetail` classes into a single `UserViewSet` class.  We can remove the two view classes, and replace them with a single ViewSet class:
```

Next we're going to replace the `SnippetList`, `SnippetDetail` and `SnippetHighlight` view classes.  We can remove the three views, and again replace them with a single class.

    from rest_framework import permissions
    from rest_framework.decorators import action
    from rest_framework.response import Response

    class SnippetViewSet(viewsets.ModelViewSet):
        """

>         This view set automatically provides `list`, `create`, `retrieve`,

Reviewer: ```suggestion
        This ViewSet automatically provides `list`, `create`, `retrieve`,
```

The URLs for custom actions by default depend on the method name itself. If you want to change the way url should be constructed, you can include `url_path` as a decorator keyword argument.

> ## Binding view sets to URLs explicitly

Reviewer: ```suggestion
## Binding ViewSets to URLs explicitly
```

The handler methods only get bound to the actions when we define the URLConf.

> To see what's going on under the hood let's first explicitly create a set of views from our view sets.

Reviewer: ```suggestion
To see what's going on under the hood let's first explicitly create a set of views from our ViewSets.
```

    from django.urls import path, include
    from rest_framework.routers import DefaultRouter

    from snippets import views

>     # Create a router and register our view sets with it.

Reviewer: ```suggestion
    # Create a router and register our ViewSets with it.
```

    router = DefaultRouter()
    router.register(r'snippets', views.SnippetViewSet, basename='snippet')
    router.register(r'users', views.UserViewSet, basename='user')

    # The API URLs are now determined automatically by the router.
    urlpatterns = [
        path('', include(router.urls)),
    ]

> Registering the view sets with the router is similar to providing a urlpattern.  We include two arguments - the URL prefix for the views, and the view set itself.

Reviewer: ```suggestion
Registering the ViewSets with the router is similar to providing a urlpattern.  We include two arguments - the URL prefix for the views, and the view set itself.
```

The `DefaultRouter` class we're using also automatically creates the API root view for us, so we can now delete the `api_root` function from our `views` module.

> ## Trade-offs between views vs view sets

Reviewer: ```suggestion
## Trade-offs between views vs ViewSets
```

> Using view sets can be a really useful abstraction.  It helps ensure that URL conventions will be consistent across your API, minimizes the amount of code you need to write, and allows you to concentrate on the interactions and representations your API provides rather than the specifics of the URL conf.

Reviewer: ```suggestion
Using ViewSets can be a really useful abstraction.  It helps ensure that URL conventions will be consistent across your API, minimizes the amount of code you need to write, and allows you to concentrate on the interactions and representations your API provides rather than the specifics of the URL conf.
```

> That doesn't mean it's always the right approach to take.  There's a similar set of trade-offs to consider as when using class-based views instead of function-based views.  Using view sets is less explicit than building your views individually.

Reviewer: ```suggestion
That doesn't mean it's always the right approach to take. There's a similar set of trade-offs to consider as when using class-based views instead of function-based views. Using ViewSets is less explicit than building your API views individually.
```
